### PR TITLE
#467 add viewmodel saved state support

### PR DESCRIPTION
**Pull request by kedzie**
This is for issue #467

### CODE REVIEW

**Comment by kedzie** [comment]
@arnaudgiuliani any thoughts?  I was wondering about the `sharedViewModel`, if we need variations of the methods for those as well.  `sharedViewModelWithState`, I was trying to avoid that.

**Comment by arnaudgiuliani** [comment]
Hello @kedzie , good job.

Could you propose something without adding a new injection function? 

We could pass a param from the `viewModel` dsl keyword like `viewModel(useState = true) { (state : ...) -> YourVIewModel(sate) }` or even propose a new keyword `stateViewModel { ... }`

Don't know what could be the best.

And from the resolution step, we can detect if we have the flag property to use the new factory

**Comment by kedzie** [comment]
Will do !  thanks for taking a look

**Comment by kedzie** [comment]
@arnaudgiuliani Great advice! made the PR much much cleaner.

**Comment by coro101** [comment]
@kedzie I hope this pull request will be merged soon! Great work!

**Comment by ustadenis** [comment]
@kedzie Could you please say whan it will be merged? Thank you.

**Comment by arnaudgiuliani** [comment]
@kedzie one last question, why do we need to expose `defaultArguments: Bundle? = null` in `FragmentExt.kt` or other viewModel() extension?

**Comment by kedzie** [comment]
The [AbstractSavedStateVMFactory](https://developer.android.com/reference/androidx/lifecycle/AbstractSavedStateVMFactory.html) takes a `Bundle defaultArgs` for initial values in the SavedStateHandle.  The optional argument allows those to be passed in.

**Comment by arnaudgiuliani** [comment]
A lazy approach can be interesting effectively. The thing is to avoid conflict with the actual parameters lambda. Could you reuse it also? 🤔 

I mean, reuse the `parametersOf` lambda stuff to pass your bundle.

**Comment by kedzie** [comment]
i can think of 2 ways to pass in the default bundle in the definition parameters:

1. pass Bundle as an additional **first** parameter of definitions.. so something like `by currentScope.viewModel(this ) { parametersOf(Bundle().apply { putString("bling" "blang") }, "vm2") }` . problem is that bundle is optional so when resolving this must detect if first parameter is a bundle?  what if 2nd parameter is also a bundle?  resolving can be complicated..
2. pass Bundle as an additional **last** parameter of definitions.. so something like `by currentScope.viewModel(this ) { parametersOf("vm2", Bundle().apply { putString("bling" "blang") }) }` . if we know the # of parameters expected would could detect if an extra one has been passed in, which is the Bundle.  
3. Could have bundle be passed into the parameters method and populated... but this doesn't really match the rest of koin.. `by currentScope.viewModel(this) { defaultArgs -> 
defaultArgs.putString("bling", "blang")
parametersOf("vm2") }`

_An argument for keeping it the way it is:_

`ViewModelParameters` already has another lambda for ` val from: ViewModelStoreOwnerDefinition? = null`
where `typealias ViewModelStoreOwnerDefinition = () -> ViewModelStoreOwner`

So the defaultArguments bundle as it's own lambda follows that same pattern.

**Comment by arnaudgiuliani** [comment]
@kedzie yeah, not so clean to have this bundle here :/ But then we need those args at the creation of the VM. Perhaps best is to keep it as an optional parameter finally?! 🤔

**Comment by kedzie** [comment]
keep it as a lazy optional parameter I think is best.

**Comment by kedzie** [comment]
I could add some espresso instrumentation tests to run through the `androidx-samples` . so we know at build time all this stuff resolves..

**Comment by jgavazzisp** [comment]
Hi team, any update in when this branch will get merged in?

Thanks

File: koin-projects/koin-androidx-viewmodel/src/main/java/org/koin/androidx/viewmodel/ViewModelResolution.kt [hunk 1]
```diff
@@ -61,9 +67,18 @@ fun <T : ViewModel> Scope.createViewModelProvider(
 ): ViewModelProvider {
     return ViewModelProvider(
             vmStore,
-            object : ViewModelProvider.Factory {
-                override fun <T : ViewModel> create(modelClass: Class<T>): T {
-                    return get(parameters.clazz, parameters.qualifier, parameters.parameters)
+            if(beanRegistry.findDefinition(parameters.qualifier, parameters.clazz)!!.isStateViewModel()) {
```
Comment (arnaudgiuliani) [inline koin-projects/koin-androidx-viewmodel/src/main/java/org/koin/androidx/viewmodel/ViewModelResolution.kt]: avoid !! here please 👍

**Comment by arnaudgiuliani** [comment]
Seems to be fine. Just one comment, to avoid using `!!` operator. @kedzie no need of expresso tests, too heavy to pass.  Good job 👍

**Pull request merged**